app/models/article.rb

```diff
@@ -1,4 +1,34 @@
 class Article < ApplicationRecord
   validates :title, presence: true
   validates :body, presence: true
+  validates :slug, uniqueness: true
+
+  before_save :set_slug
```

Comment:
Since you only set_slug to be run before_save if the title has actually changed, you can call it like this:

before_save :set_slug, if: -> { self.title_changed? }

Which is a good consideration to have to avoid unnecessary state updates on the model.